native/cocos/2d/renderer/Batcher2d.cpp

```diff
@@ -0,0 +1,258 @@
+                }
+            }
+        }
+
+        if (curNode->getChangedFlags()) {
+            fillVertexBuffers(entity);
+        }
+        fillIndexBuffers(entity);
+    }
+
+    auto& children = curNode->getChildren();
+    for (index_t i = children.size() - 1; i >= 0; i--) {
+        assert(length < 1000000);
+        nodeStack[length++] = children[i];
+    }
+}
+}
+
+void Batcher2d::generateBatch(RenderDrawInfo* entity) {
+    if (entity == nullptr) {
+        return;
+    }
+    // if (entity->getIsMeshBuffer()) {
+    //     // Todo MeshBuffer RenderData
+    // } else {
+    UIMeshBuffer* currMeshBuffer = entity->getMeshBuffer();
+
+    currMeshBuffer->setDirty(true);
+
+    auto* ia = currMeshBuffer->requireFreeIA(getDevice());
+    uint32_t indexCount = currMeshBuffer->getIndexOffset() - _indexStart;
+    if (ia == nullptr) {
+        return;
+    }
+
+    ia->setFirstIndex(_indexStart);
+    ia->setIndexCount(indexCount);
+    _indexStart = currMeshBuffer->getIndexOffset();
+    //}
+
+    this->_currBID = -1;
+
+    // stencilstage
+
+    // Todo blendState & stencil State
+    auto* curdrawBatch = _drawBatchPool.alloc();
+    curdrawBatch->setVisFlags(_currLayer);
+    // curdrawBatch
+    curdrawBatch->setInputAssembler(ia);
+    curdrawBatch->setUseLocalFlag(nullptr); // todo usLocal
+    curdrawBatch->fillPass(_currMaterial, nullptr, 0, nullptr, 0);
+    auto& _pass = curdrawBatch->getPasses().at(0);
+
+    curdrawBatch->setDescriptorSet(getDescriptorSet(_currTexture, _currSampler, _pass->getLocalSetLayout()));
+    _batches.push_back(curdrawBatch);
+}
+
+void Batcher2d::resetRenderStates() {
+    _currMaterial = nullptr;
+    _currTexture = nullptr;
+    _currTextureHash = 0;
+    _currSampler = nullptr;
+    _currSamplerHash = 0;
+    _currLayer = 0;
+    _currEntity = nullptr;
+}
+
+gfx::DescriptorSet* Batcher2d::getDescriptorSet(gfx::Texture* texture, gfx::Sampler* sampler, gfx::DescriptorSetLayout* _dsLayout) {
+    ccstd::hash_t hash = 2;
+    if (texture != nullptr) {
+        ccstd::hash_combine(hash, texture->getHash());
+    }
+    if (sampler != nullptr) {
+        ccstd::hash_combine(hash, sampler->getHash());
+    }
+    auto iter = _descriptorSetCache.find(hash);
+    if (iter != _descriptorSetCache.end()) {
+        return iter->second;
+    }
+    _dsInfo.layout = _dsLayout;
+    auto ds = getDevice()->createDescriptorSet(_dsInfo);
+
+    ds->bindTexture(static_cast<uint>(pipeline::ModelLocalBindings::SAMPLER_SPRITE), texture);
+    ds->bindSampler(static_cast<uint>(pipeline::ModelLocalBindings::SAMPLER_SPRITE), sampler);
+    ds->update();
+
+    _descriptorSetCache.emplace(hash, ds);
+
+    return ds;
+}
+
+// update vertex code
+
+bool Batcher2d::initialize() {
+    return true;
+}
+
+void Batcher2d::update() {
+    updateVertDirtyRenderer();
+    fillBuffersAndMergeBatches();
+    resetRenderStates();
+
+    for (const auto scene : Root::getInstance()->getScenes()) {
```

Comment:
Should use pointer instead
```
const auto *scene
```

Reply:
auto == IntrusivePtr
auto* will cause a compile error